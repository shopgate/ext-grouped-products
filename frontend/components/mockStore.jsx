import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import mockRenderOptions from '@shopgate/pwa-common/helpers/mocks/mockRenderOptions';

/**
 * Creates a component instance which is wrapped by a store.
 * @param {JSX} Component The components to render.
 * @param {Object} state A mocked state.
 * @param {Object} [props={}] Component props.
 * @param {Function} [dispatch=null] An optional replacement for the dispatch method.
 * @return {ReactWrapper}
 */
export const createWrappedComponent = (Component, state, props = {}, dispatch = null) => {
  const store = configureStore([thunk])(state);

  if (dispatch) {
    store.dispatch = dispatch;
  }

  return mount(
    <Provider store={store}>
      <Component {...props} />
    </Provider>,
    mockRenderOptions
  );
};
