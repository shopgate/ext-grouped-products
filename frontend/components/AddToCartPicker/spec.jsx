import React, { Component } from 'react';
import { createWrappedComponent } from '../mockStore';
import {
  mockedState,
  mockedProduct,
  mockedIgnoredQuantityProduct,
  mockedNotOrderableProduct,
} from '../mock';

// eslint-disable-next-line require-jsdoc
class mockedAdToCartButton extends Component {
  // eslint-disable-next-line require-jsdoc
  static get propTypes() {
    return {};
  }

  // eslint-disable-next-line require-jsdoc
  constructor(props) {
    super(props);
    this.state = {
      // eslint-disable-next-line react/no-unused-state
      showCheckMark: false,
    };
  }

  // eslint-disable-next-line require-jsdoc
  render() {
    return <div><button onClick={this.handleClick}>Add to cart</button></div>;
  }
}
// Missing css-spring mock on pwa-ui-shared
jest.mock('@shopgate/pwa-ui-shared/AddToCartButton', () => mockedAdToCartButton);

jest.mock('@shopgate/pwa-ui-shared/AddToCartButton/style', () => (
  {
    buttonSize: 10,
    iconSize: 10,
    buttonWrapper: () => { },
    buttonWrapperNoShadow: () => { },
  }
));

jest.mock('../config', () => { }, { virtual: true });

jest.mock('@shopgate/pwa-common-commerce/cart/actions/addProductsToCart', () =>
  jest.fn().mockReturnValue('mocked_add_products_to_cart_action'));

jest.mock('../../config', () => ({
  maxQuantityPickerEntries: 5,
}), { virtual: true });

jest.mock('react-portal', () => (
  ({ isOpened, children }) => (
    isOpened ? children : null
  )
));
try {
  jest.doMock('@shopgate/pwa-common/node_modules/react-portal', () => (
    ({ isOpened, children }) => (
      isOpened ? children : null
    )
  ));
} catch (e) {
  // Do nothing.
}

jest.mock('@shopgate/pwa-ui-shared/Sheet', () => ({ children }) => children);

const mockHandleAddToCart = jest.fn();

jest.useFakeTimers();

describe('<AddToCartPicker />', () => {
  // eslint-disable-next-line global-require
  const AddToCartPicker = require('./index.jsx').default;

  /**
   * Creates a component with a provided store state.
   * @param {Object} product A product to set the component props for.
   * @param {Object} [buttonProps={}] Mocked button props.
   * @return {ReactWrapper}
   */
  const createComponent = (product, buttonProps = {}) => {
    const { id, stock } = product;

    const mockedProps = {
      buttonProps,
      productId: id,
      handleAddToCart: mockHandleAddToCart,
      stock,
      onClose: () => { },
    };

    return createWrappedComponent(AddToCartPicker, mockedState, mockedProps);
  };

  it('should create picker items like expected when quantity is not ignored', () => {
    const component = createComponent(mockedProduct);
    const picker = component.find('Picker').first();
    const pickerItems = picker.prop('items');
    expect(pickerItems).toHaveLength(4);
    expect(pickerItems[0]).toEqual({
      label: '1',
      value: 1,
    });
    expect(pickerItems[pickerItems.length - 1]).toEqual({
      label: '4',
      value: 4,
    });
  });

  it('should create picker items like expected when quantity is ignored', () => {
    const component = createComponent(mockedIgnoredQuantityProduct);
    const picker = component.find('Picker').first();
    const pickerItems = picker.prop('items');
    expect(pickerItems).toHaveLength(5);
    expect(pickerItems[0]).toEqual({
      label: '1',
      value: 1,
    });
    expect(pickerItems[pickerItems.length - 1]).toEqual({
      label: '5',
      value: 5,
    });
  });

  it('should update the component when expected', () => {
    const component = createComponent(mockedProduct);
    const picker = component.find('AddToCartPicker').first();

    const props = picker.props();
    const state = { ...picker.instance().state };

    const scu = picker.instance().shouldComponentUpdate.bind(picker.instance());

    // Nothing changed
    expect(scu({ ...props }, { ...state })).toBe(false);

    // Some ignored prop changed
    expect(scu({
      ...props,
      clickDelay: 200,
    }, { ...state })).toBe(false);

    // Button props changed
    expect(scu({
      ...props,
      buttonProps: {
        ...props.buttonProps,
        isLoading: true,
      },
    }, { ...state })).toBe(true);

    // State changed
    expect(scu({
      ...props,
    }, {
      ...state,
      addedQuantity: 3,
    })).toBe(true);
  });
});
