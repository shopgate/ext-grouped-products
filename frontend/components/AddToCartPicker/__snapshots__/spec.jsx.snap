// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddToCartPicker /> should render the component as expected 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <AddToCartPicker
    buttonProps={Object {}}
    clickDelay={150}
    handleAddToCart={[MockFunction]}
    productId="1337"
    stock={
      Object {
        "ignoreQuantity": false,
        "info": "Info",
        "maxOrderQuantity": 1000,
        "minOrderQuantity": 1,
        "orderable": true,
        "quantity": 4,
      }
    }
  >
    <Picker
      buttonComponent={[Function]}
      buttonProps={
        Object {
          "addedQuantity": 0,
          "className": null,
          "handleAddToCart": [Function],
          "hasLoading": true,
          "isDisabled": false,
          "isLoading": false,
          "isOrderable": true,
          "noShadow": false,
          "onClick": [Function],
          "type": "default",
        }
      }
      className="css-ibd9un"
      disabled={false}
      duration={300}
      isOpen={false}
      items={
        Array [
          Object {
            "label": "1",
            "value": 1,
          },
          Object {
            "label": "2",
            "value": 2,
          },
          Object {
            "label": "3",
            "value": 3,
          },
          Object {
            "label": "4",
            "value": 4,
          },
        ]
      }
      label=""
      listComponent={[Function]}
      modalComponent={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onClose={[Function]}
      onSelect={[Function]}
      placeholder="Pick ..."
      value={null}
    >
      <div
        className="css-ibd9un"
      >
        <PickerAddToCartButton
          addedQuantity={0}
          className={null}
          disabled={false}
          handleAddToCart={[Function]}
          hasLoading={true}
          isDisabled={false}
          isLoading={false}
          isOrderable={true}
          label="Pick ..."
          noShadow={false}
          onClick={[Function]}
          openList={[Function]}
          type="default"
          value=""
        >
          <button
            data-test-id="addToCartButton"
            onClick={[Function]}
          />
        </PickerAddToCartButton>
        <Component
          isOpened={true}
          onClose={[Function]}
        >
          <Component
            duration={300}
            isOpen={false}
            onClose={[Function]}
          >
            <Component
              duration={300}
              isOpen={false}
              onClose={[Function]}
              title="grouped_products.add_to_cart_picker.headline"
            >
              <Component
                items={
                  Array [
                    Object {
                      "disabled": false,
                      "label": "1",
                      "value": 1,
                    },
                    Object {
                      "disabled": false,
                      "label": "2",
                      "value": 2,
                    },
                    Object {
                      "disabled": false,
                      "label": "3",
                      "value": 3,
                    },
                    Object {
                      "disabled": false,
                      "label": "4",
                      "value": 4,
                    },
                  ]
                }
                onSelect={[Function]}
                selectedIndex={-1}
              >
                <List>
                  <List
                    className=""
                  >
                    <ul
                      className=""
                    >
                      <ListItem
                        className="css-17flzkh"
                        isSelected={false}
                        key="child-0/.$1"
                      >
                        <li
                          className="css-1p5q8vq css-17flzkh css-16gj3o1"
                          data-test-id="listItem"
                        >
                          <div
                            className="css-1nnz83b"
                          >
                            <Item
                              className={null}
                              image={null}
                              isDisabled={false}
                              isSelected={false}
                              key="1"
                              link={null}
                              onClick={[Function]}
                              rightComponent={null}
                              testId={null}
                              title="1"
                            >
                              <div
                                aria-hidden={true}
                                data-test-id={null}
                                onClick={[Function]}
                              >
                                <Glow
                                  className={null}
                                  color="#f7f7f7"
                                  styles={
                                    Object {
                                      "container": null,
                                      "glow": null,
                                    }
                                  }
                                >
                                  <div
                                    aria-hidden={true}
                                    className="css-13azwyo"
                                    onClick={[Function]}
                                    style={null}
                                  >
                                    <div
                                      className="css-1bmd5dz"
                                      style={Object {}}
                                    />
                                    <Grid
                                      className="css-1qcr7hz"
                                      component="div"
                                      wrap={false}
                                    >
                                      <div
                                        className="css-1qcr7hz css-h7r55a"
                                      >
                                        <div
                                          className="css-jwxytc"
                                        />
                                        <GridItem
                                          className="css-10fxwvh"
                                          component="div"
                                          grow={1}
                                          shrink={1}
                                        >
                                          <div
                                            className="css-10fxwvh css-1gs0ko2"
                                          >
                                            1
                                          </div>
                                        </GridItem>
                                      </div>
                                    </Grid>
                                  </div>
                                </Glow>
                              </div>
                            </Item>
                          </div>
                        </li>
                      </ListItem>
                      <ListItem
                        className="css-17flzkh"
                        isSelected={false}
                        key="child-1/.$2"
                      >
                        <li
                          className="css-1p5q8vq css-17flzkh css-16gj3o1"
                          data-test-id="listItem"
                        >
                          <div
                            className="css-1nnz83b"
                          >
                            <Item
                              className={null}
                              image={null}
                              isDisabled={false}
                              isSelected={false}
                              key="2"
                              link={null}
                              onClick={[Function]}
                              rightComponent={null}
                              testId={null}
                              title="2"
                            >
                              <div
                                aria-hidden={true}
                                data-test-id={null}
                                onClick={[Function]}
                              >
                                <Glow
                                  className={null}
                                  color="#f7f7f7"
                                  styles={
                                    Object {
                                      "container": null,
                                      "glow": null,
                                    }
                                  }
                                >
                                  <div
                                    aria-hidden={true}
                                    className="css-13azwyo"
                                    onClick={[Function]}
                                    style={null}
                                  >
                                    <div
                                      className="css-1bmd5dz"
                                      style={Object {}}
                                    />
                                    <Grid
                                      className="css-1qcr7hz"
                                      component="div"
                                      wrap={false}
                                    >
                                      <div
                                        className="css-1qcr7hz css-h7r55a"
                                      >
                                        <div
                                          className="css-jwxytc"
                                        />
                                        <GridItem
                                          className="css-10fxwvh"
                                          component="div"
                                          grow={1}
                                          shrink={1}
                                        >
                                          <div
                                            className="css-10fxwvh css-1gs0ko2"
                                          >
                                            2
                                          </div>
                                        </GridItem>
                                      </div>
                                    </Grid>
                                  </div>
                                </Glow>
                              </div>
                            </Item>
                          </div>
                        </li>
                      </ListItem>
                      <ListItem
                        className="css-17flzkh"
                        isSelected={false}
                        key="child-2/.$3"
                      >
                        <li
                          className="css-1p5q8vq css-17flzkh css-16gj3o1"
                          data-test-id="listItem"
                        >
                          <div
                            className="css-1nnz83b"
                          >
                            <Item
                              className={null}
                              image={null}
                              isDisabled={false}
                              isSelected={false}
                              key="3"
                              link={null}
                              onClick={[Function]}
                              rightComponent={null}
                              testId={null}
                              title="3"
                            >
                              <div
                                aria-hidden={true}
                                data-test-id={null}
                                onClick={[Function]}
                              >
                                <Glow
                                  className={null}
                                  color="#f7f7f7"
                                  styles={
                                    Object {
                                      "container": null,
                                      "glow": null,
                                    }
                                  }
                                >
                                  <div
                                    aria-hidden={true}
                                    className="css-13azwyo"
                                    onClick={[Function]}
                                    style={null}
                                  >
                                    <div
                                      className="css-1bmd5dz"
                                      style={Object {}}
                                    />
                                    <Grid
                                      className="css-1qcr7hz"
                                      component="div"
                                      wrap={false}
                                    >
                                      <div
                                        className="css-1qcr7hz css-h7r55a"
                                      >
                                        <div
                                          className="css-jwxytc"
                                        />
                                        <GridItem
                                          className="css-10fxwvh"
                                          component="div"
                                          grow={1}
                                          shrink={1}
                                        >
                                          <div
                                            className="css-10fxwvh css-1gs0ko2"
                                          >
                                            3
                                          </div>
                                        </GridItem>
                                      </div>
                                    </Grid>
                                  </div>
                                </Glow>
                              </div>
                            </Item>
                          </div>
                        </li>
                      </ListItem>
                      <ListItem
                        className="css-17flzkh"
                        isSelected={false}
                        key="child-3/.$4"
                      >
                        <li
                          className="css-1p5q8vq css-17flzkh css-16gj3o1"
                          data-test-id="listItem"
                        >
                          <div
                            className="css-1nnz83b"
                          >
                            <Item
                              className={null}
                              image={null}
                              isDisabled={false}
                              isSelected={false}
                              key="4"
                              link={null}
                              onClick={[Function]}
                              rightComponent={null}
                              testId={null}
                              title="4"
                            >
                              <div
                                aria-hidden={true}
                                data-test-id={null}
                                onClick={[Function]}
                              >
                                <Glow
                                  className={null}
                                  color="#f7f7f7"
                                  styles={
                                    Object {
                                      "container": null,
                                      "glow": null,
                                    }
                                  }
                                >
                                  <div
                                    aria-hidden={true}
                                    className="css-13azwyo"
                                    onClick={[Function]}
                                    style={null}
                                  >
                                    <div
                                      className="css-1bmd5dz"
                                      style={Object {}}
                                    />
                                    <Grid
                                      className="css-1qcr7hz"
                                      component="div"
                                      wrap={false}
                                    >
                                      <div
                                        className="css-1qcr7hz css-h7r55a"
                                      >
                                        <div
                                          className="css-jwxytc"
                                        />
                                        <GridItem
                                          className="css-10fxwvh"
                                          component="div"
                                          grow={1}
                                          shrink={1}
                                        >
                                          <div
                                            className="css-10fxwvh css-1gs0ko2"
                                          >
                                            4
                                          </div>
                                        </GridItem>
                                      </div>
                                    </Grid>
                                  </div>
                                </Glow>
                              </div>
                            </Item>
                          </div>
                        </li>
                      </ListItem>
                    </ul>
                  </List>
                </List>
              </Component>
            </Component>
          </Component>
        </Component>
      </div>
    </Picker>
  </AddToCartPicker>
</Provider>
`;
